bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -461,87 +489,343 @@ public Boolean isEnabled(String ruleUID) {
+        }
+    }
 
-                return resolvedRule;
+    /**
+     * Updates the content of the {@link Map} that maps the template to rules, using it to complete their definitions.
+     *
+     * @param templateUID the {@link RuleTemplate}'s UID specifying the template.
+     * @param ruleUID the {@link Rule}'s UID specifying a rule created by the specified template.
+     * @param resolved specifies if the {@link Map} should be updated by adding or removing the specified rule
+     *            accordingly if the rule is resolved or not.
+     */
+    private void updateRuleTemplateMaping(String templateUID, String ruleUID, boolean resolved) {
+        synchronized (this) {
+            Set<String> ruleUIDs = mapTemplateToRules.get(templateUID);
+            if (ruleUIDs == null) {
+                ruleUIDs = new HashSet<String>();
+                mapTemplateToRules.put(templateUID, ruleUIDs);
+            }
+            if (resolved) {
+                ruleUIDs.remove(ruleUID);
+            } else {
+                ruleUIDs.add(ruleUID);
             }
         }
-        return rule;
     }
 
     @Override
     protected void addProvider(Provider<Rule> provider) {
         super.addProvider(provider);
         Collection<Rule> rules = new LinkedList<Rule>(elementMap.get(provider));
         for (Rule rule : rules) {
-            updateRuleByTemplate(provider, rule);
+            try {
+                Rule resolvedRule = resolveRuleByTemplate(rule);
+                if (rule != resolvedRule) {
+                    update(resolvedRule);
+                }
+            } catch (Exception e) {
+                logger.error("Added rule '{}' is invalid", rule.getUID(), e);
+            }
         }
     }
 
     @Override
     public void added(Provider<Rule> provider, Rule element) {
-        super.added(provider, element);
-        updateRuleByTemplate(provider, element);
-    }
+        Rule ruleCopy = RuleUtils.getRuleCopy(element);
+        super.added(provider, ruleCopy);
+        postRuleAddedEvent(ruleCopy);
+        try {
+            Rule resolvedRule = resolveRuleByTemplate(ruleCopy);
+            if (resolvedRule.getTemplateUID() == null) {
+                RuleStatusInfo statusInfo = disabledRulesStorage != null
+                        && disabledRulesStorage.get(ruleCopy.getUID()) == null
+                                ? new RuleStatusInfo(RuleStatus.INITIALIZING)
+                                : new RuleStatusInfo(RuleStatus.DISABLED);
+                addToRuleEngine(resolvedRule, statusInfo);
 
-    private void updateRuleByTemplate(Provider<Rule> provider, Rule rule) {
-        Rule resolvedRule = resolveRuleByTemplate(rule);
-        if (rule != resolvedRule) {
-            if (update(resolvedRule) == null) {
-                super.updated(provider, rule, resolvedRule);
             }
+            if (ruleCopy != resolvedRule) {
+                update(resolvedRule);
+            }
+        } catch (Exception e) {
```

Comment:
We should not catch generic Exceptions. Can you make this more specific?

Reply:
OK